packages/flutter_tools/lib/src/ios/devices.dart

```diff
@@ -887,12 +1068,110 @@ class IOSDeviceLogReader extends DeviceLogReader {
 
   static const int minimumUniversalLoggingSdkVersion = 13;
 
+  /// Use `idevicesyslog` to stream logs from the device when one of the
+  /// following criteria is met:
+  ///
+  /// 1) The device is a physically attached CoreDevice.
+  /// 2) The device has iOS 16 or greater and it's being debugged from CI.
+  /// 3) The device has iOS 12 or lower.
+  ///
+  /// Syslog stopped working on iOS 13 (https://github.com/flutter/flutter/issues/41133).
+  /// However, from at least iOS 16, it has began working again. It's unclear
+  /// why it started working again.
+  @visibleForTesting
+  bool get useSyslogLogging {
+    // `ios-deploy` stopped working with iOS 17 / Xcode 15, so use `idevicesyslog` instead.
+    // However, `idevicesyslog` does not work with iOS 17 wireless devices.
+    if (_isCoreDevice && !_isWirelesslyConnected) {
+      return true;
+    }
+
+    // Use both logs from `idevicesyslog` and `ios-deploy` when debugging from CI system
+    // since sometimes `ios-deploy` does not return the device logs:
+    // https://github.com/flutter/flutter/issues/121231
+    if (_usingCISystem && _majorSdkVersion >= 16) {
+      return true;
+    }
+    if (_majorSdkVersion < minimumUniversalLoggingSdkVersion) {
+      return true;
+    }
+    return false;
+  }
+
+  /// Use the Dart VM to stream logs from the device when one of the following
+  /// criteria is met:
+  ///
+  /// 1) The device is a CoreDevice and wirelessly connected.
+  /// 2) The device has iOS 13 or greater and [_iosDeployDebugger] is null or
+  /// the [_iosDeployDebugger] debugger is not attached.
+  ///
+  /// This value may change if [_iosDeployDebugger] changes.
+  @visibleForTesting
+  bool get useUnifiedLogging {
+    // `idevicesyslog` doesn't work on wireless devices, so use logs from Dart VM instead.
+    if (_isCoreDevice) {
+      if (_isWirelesslyConnected) {
+        return true;
+      }
+      return false;
+    }
+
+    // Prefer the more complete logs from the attached debugger.
+    if (_majorSdkVersion >= minimumUniversalLoggingSdkVersion && (_iosDeployDebugger == null || !_iosDeployDebugger!.debuggerAttached)) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /// Determine whether to listen to the Dart VM for logging events. Returns
+  /// true when one of the following criteria is met:
+  ///
+  /// 1) The device is a CoreDevice and wirelessly connected.
+  /// 2) The device has iOS 13 or greater.
```

Comment:
For the 3 functions here - i think it's pretty easy to mess up the logic and we have to be very very careful here. Besides writing extensive unit test, one idea is to combine all of them into 1 function, and return an enum case. This way we co-locate these related logics.

Reply:
How would an enum work when we sometimes use multiple at a time?